fetch.bs

```diff
@@ -3630,51 +3611,75 @@ steps:
 
-  <p class=note>This operates on <var>response</var> as this algorithm is not supposed to observe
-  <var>internalResponse</var>. That would allow an attacker to use hashes as an oracle.
 
- <li>
-  <p>If <var>fetchParams</var>'s <a for="fetch params">process response</a> is non-null, then:
+   <li><p>If <var>request</var>'s <a for=request>response tainting</a> is "<code>opaque</code>",
+   <var>response</var> is a <a>network error</a>, or <var>response</var>'s <a for=response>body</a>
+   is null, then run <var>processBodyError</var> and abort these steps.
 
-  <ol>
-   <li><p>Let <var>processResponse</var> be this step: run <var>fetchParams</var>'s
-   <a for="fetch params">process response</a> given <var>response</var>.
+   <li>
+    <p>Let <var>processBody</var> given <var>bytes</var> be these steps:
 
-   <li><p><a>Queue a fetch task</a> given <var>processResponse</var> and <var>fetchParams</var>'s
-   <a for="fetch params">task destination</a>.
+    <ol>
+     <li><p>If <var>bytes</var> do not
+     <a href=https://w3c.github.io/webappsec-subresource-integrity/#does-response-match-metadatalist>match</a>
+     <var>request</var>'s <a for=request>integrity metadata</a>, then run
+     <var>processBodyError</var> and abort these steps. [[!SRI]]
+
+     <li><p>Set <var>response</var>'s <a for=response>body</a> to the first return value of
+     <a for=BodyInit>safely extracting</a> <var>bytes</var>.
+
+     <li><p>Run <a>fetch finale</a> given <var>fetchParams</var> and <var>response</var>.
+    </ol>
+
+   <li><p><a for=body>Fully read</a> <var>response</var>'s <a for=response>body</a> given
+   <var>processBody</var> and <var>processBodyError</var>.
   </ol>
 
- <li><p><a lt=wait for=body>Wait</a> for <var>internalResponse</var>'s
- <a for=response>body</a>.
+ <li><p>Otherwise, run <a>fetch finale</a> given <var>fetchParams</var> and <var>response</var>.
+</ol>
+
+<hr>
+
+<p>The <dfn>fetch finale</dfn>, given a <a for=/>fetch params</a> <var>fetchParams</var> and a
+<a for=/>response</a> <var>response</var>, run these steps:
+
+<ol>
+ <li><p>If <var>fetchParams</var>'s <a for="fetch params">process response</a> is non-null, then
+ <a>queue a fetch task</a> to run <var>fetchParams</var>'s
+ <a for="fetch params">process response</a> given <var>response</var>, with <var>fetchParams</var>'s
+ <a for="fetch params">task destination</a>.
 
  <li>
-  <p>Let <var>doneAlgorithm</var> be these steps:
+  <p>If <var>fetchParams</var>'s <a for="fetch params">process response end-of-body</a> is non-null,
+  then:
 
   <ol>
-   <li><p>Set <var>request</var>'s <a>done flag</a>.
+   <li><p>Let <var>processBody</var> given <var>nullOrFailureOrBytes</var> be this step: run
+   <var>fetchParams</var>'s <a for="fetch params">process response end-of-body</a> given
+   <var>response</var> and <var>nullOrFailureOrBytes</var>.
+
+   <li><p>Let <var>processBodyError</var> be to run <var>processBody</var> given failure.
```

Comment:
Although this is clever, I think it'd be easier to understand if it directly called to the output algorithm, i.e.

> Let _processBodyError_ be this step: run _fetchParam_'s process response end-of-body given _response_ and failure.